asrtoolkit/alignment/align.py

```diff
@@ -0,0 +1,181 @@
+    :return: List of Extractors objects to use for text-text alignment
+    note: ngram extractors below filter out stopwords and number words/symbols
+    """
+    noun_chunk_extractor = Extractor(
+        lambda doc: list(filter(lambda x: len(x) > 3, list(noun_chunks(doc)))))
+    tetragram_extractor = Extractor(
+        lambda doc: list(ngrams(doc, 4, filter_stops=True, filter_nums=True)))
+    trigram_extractor = Extractor(
+        lambda doc: list(ngrams(doc, 3, filter_stops=True, filter_nums=True)))
+    bigram_extractor = Extractor(lambda doc: list(
+        ngrams(doc, 2, filter_stops=False, filter_nums=False)))
+    unigram_extractor = Extractor(lambda doc: list(
+        ngrams(doc, 1, filter_stops=False, filter_nums=False)))
+
+    extractor_list = [
+        noun_chunk_extractor,
+        tetragram_extractor,
+        trigram_extractor,
+        bigram_extractor,
+    ]
+
+    if use_unigrams:
+        extractor_list.append(unigram_extractor)
+
+    return extractor_list
+
+
+def lattices_to_aligned_doc(time_annotated_tokens, reference_tokens,
+                            extractors, num_extractions):
+    """
+    From two Lists of Word-Dicts, returns AlignedDoc (already aligned)
+    :param time_annotated_tokens: List[Dict]; dicts with token-specific data
+      including, at minimum, token start time and duration
+    :param reference_tokens: List[Dict]; dicts with token-specific data
+      including speaker(id) and gender
+    :return: AlignedDoc object; contains
+      end-product of text-text alignment and
+      methods to retrieve matches with and without token-level metadata
+    """
+    # Create Spacy Docs from List of Tokens + Metadata
+    hyp = init_spacy_document(lattice=time_annotated_tokens,
+                              token_key="gk_token")
+    ref = init_spacy_document(lattice=reference_tokens, token_key="token")
+    return AlignedDoc(hyp, ref, extractors, num_extractions)
+
+
+def get_segments_from_alignments(
+    aligned_token_lattice,
+    reference_doc,
+    token_idx="token_idx",
+    max_duration=15,
+):
+    """
+    Uses Aligned Tokens to Segment Reference Text into STM Lines
+    :param aligned_token_lattice: List of Dicts,STM
+      each containing word-level information;
+      all tokens are paired and contain data requisite for STM file
+    :param max_duration: int; maximum duration of STM line audio
+    :return: segments: List[Segment],
+      each Segment object contains data for one STM Line
+    """
+    line_dict = word_lattice_to_lines(word_lattice=aligned_token_lattice,
+                                      MAX_DURATION=max_duration)
+    segments = dict_to_segments(line_dict=line_dict,
+                                doc=reference_doc,
+                                token_idx=token_idx)
+
+    return line_dict, segments
+
+
+def align(gk_json, ref_tokens, num_extractions=4, max_duration=15):
+    """
+    Given a gk_json and ref_tokens, apply extractors and align
```

Comment:
Does this function have to take gk JSON or could it use any ASR toolkit transcript object (I think they're called time_aligned_somethings...)? It would be nice to be more general if possible.

Reply:
Right now it does because we require word-level hypotheses with time information. We may revisit if another data format contains that information.